### PR TITLE
Add KB09-01 keyboard

### PR DESCRIPTION
**Pull request by DOIO2022**
## Description

ADD KB09-01 keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/doio/kb09/keyboard.json [hunk 1]
```diff
@@ -0,0 +1,126 @@
+{
+    "manufacturer": "DOIO",
+    "keyboard_name": "KB09-01",
+    "maintainer": "DOIO2022",
+    "bootloader": "stm32duino",
+    "diode_direction": "COL2ROW",
+    "dynamic_keymap": {
+        "layer_count": 4
+    },
```
Comment (waffle87) [inline keyboards/doio/kb09/keyboard.json:9]: Default behaviour.
```suggestion
```

File: keyboards/doio/kb09/readme.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# doio/kb09
+
+![kb09-01](https://imgur.com/MVKeweC.png)
```
Comment (waffle87) [inline keyboards/doio/kb09/readme.md:3]:
```suggestion
![kb09-01](https://i.imgur.com/MVKeweC.png)
```

**Review by waffle87 — APPROVED** [review]

File: keyboards/doio/kb09/keymaps/via/keymap.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+enum layer_names {
+    _LAY0,
+    _LAY1,
+    _LAY2,
+    _LAY3
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [_LAY0] = LAYOUT(
```
Comment (drashna) [inline keyboards/doio/kb09/keymaps/via/keymap.c:29]: This doesn't look correct.

In fact, this looks to be much more appropriate to the controller. 
```c
         KC_UP,         KC_1,    KC_2,    KC_3, 
    KC_LEFT, KC_RGHT,   KC_4,    KC_5,    KC_6,    KC_MPLY, KC_MUTE,
         KC_DOWN,       KC_7,    KC_8,    KC_9, 
     TO(1), LWIN(KC_D))
```

File: keyboards/doio/kb09/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/doio/kb09/keymaps/default/keymap.c:1]: Nit:  please replace the tabs with spaces.

File: keyboards/doio/kb09/keymaps/via/keymap.c
Comment (drashna) [inline keyboards/doio/kb09/keymaps/via/keymap.c:1]: Nit:  please replace the tabs with spaces.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb09/keymaps/via/keymap.c (hunk 3 above)
Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/via/keymap.c:29]: HI, thank you for your suggestion,I've changed it.

File: keyboards/doio/kb09/keymaps/default/keymap.c
Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/default/keymap.c:1]: Hi , I've changed it.  Thank you!

File: keyboards/doio/kb09/keymaps/via/keymap.c
Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/via/keymap.c:1]: Hi , I've changed it. Thank you!

File: keyboards/doio/kb09/keymaps/via/keymap.c (hunk 3 above)
Comment (drashna) [reply keyboards/doio/kb09/keymaps/via/keymap.c:29]: Just a heads up, the LAYOUT config in the json needs to be updated to match, too.

Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/via/keymap.c:29]: @drashna  Hi, I've changed the keyboard.json. Thank you!

Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/via/keymap.c:29]: @drashna  Hi，Please review my changed, Thank you!

File: keyboards/doio/kb09/keymaps/via/keymap.c [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  [_LAY2] = LAYOUT(
+				KC_TRNS,		S(KC_I),	A(KC_EQL),	C(S(KC_M)),	KC_TRNS,	KC_TRNS,
+		KC_TRNS,	KC_TRNS,	S(KC_O),	A(KC_MINS), S(KC_M),
+				KC_TRNS,		KC_C,		KC_V,		C(KC_M),
+		TO(3),	KC_TRNS),
+  [_LAY3] = LAYOUT(
+				KC_TRNS,		RGB_SPI,	RGB_SPD,	RGB_HUI,	KC_TRNS,	KC_TRNS,
+		KC_TRNS,	KC_TRNS,	RGB_SAI,	RGB_SAD,	RGB_HUD,
+				KC_TRNS,		RGB_MOD,	RGB_VAI,	RGB_VAD,
+		TO(0),	KC_TRNS)
```
Comment (drashna) [inline keyboards/doio/kb09/keymaps/via/keymap.c:48]: Same here, tabs -> spaces.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb09/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  [_LAY2] = LAYOUT(
+				KC_TRNS,		S(KC_I),	A(KC_EQL),	C(S(KC_M)),	KC_TRNS,	KC_TRNS,
+		KC_TRNS,	KC_TRNS,	S(KC_O),	A(KC_MINS), S(KC_M),
+				KC_TRNS,		KC_C,		KC_V,		C(KC_M),
+		TO(3),	KC_TRNS),
+  [_LAY3] = LAYOUT(
+				KC_TRNS,		RGB_SPI,	RGB_SPD,	RGB_HUI,	KC_TRNS,	KC_TRNS,
+		KC_TRNS,	KC_TRNS,	RGB_SAI,	RGB_SAD,	RGB_HUD,
+				KC_TRNS,		RGB_MOD,	RGB_VAI,	RGB_VAD,
+		TO(0),	KC_TRNS)
```
[the remark this answers is not in the document]
Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/default/keymap.c]: HI, I changed it, thank you for your correction，and I will correct my grammar writing in the future.

File: keyboards/doio/kb09/keymaps/via/keymap.c (hunk 4 above)
Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/via/keymap.c]: HI, I changed it, thank you for your correction，and I will correct my grammar writing in the future.

Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/via/keymap.c]: @drashna  HI, I've changed it, please review again,thank you so much!

**Comment by DOIO2022** [comment]
@drashna  Hi, is there any problems? Could you review my changed? Thank you very much!

File: keyboards/doio/kb09/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	[_LAY2] = LAYOUT(
+				KC_TRNS,		S(KC_I),	A(KC_EQL),	C(S(KC_M)),	KC_TRNS,	KC_TRNS,
+		KC_TRNS,	KC_TRNS,	S(KC_O),	A(KC_MINS),	S(KC_M),
+				KC_TRNS,		KC_C,		KC_V,		C(KC_M),
+		TO(3),	KC_TRNS),
+	[_LAY3] = LAYOUT(
+				KC_TRNS,		RGB_SPI,	RGB_SPD,	RGB_HUI,	KC_TRNS,	KC_TRNS,
+		KC_TRNS,	KC_TRNS,	RGB_SAI,	RGB_SAD,	RGB_HUD,
+				KC_TRNS,		RGB_MOD,	RGB_VAI,	RGB_VAD,
+		TO(0),	KC_TRNS)
```
Comment (drashna) [inline keyboards/doio/kb09/keymaps/default/keymap.c:48]: Please replace all of these tabs with spaces.

**Review by drashna — CHANGES_REQUESTED** [review]
As of August 26, 2024, `qmk/qmk_firmware` is no longer accepting VIA-enabled keymaps as these have now transitioned to a repository under the VIA team's control.

As you've submitted a PR containing `via` or VIA-enabled keymap(s), this is your notice that they should be removed from this PR. You should now submit a secondary PR to the [VIA QMK Userspace](https://github.com/the-via/qmk_userspace_via) repository with your associated `via` or VIA-enabled keymaps instead.

File: keyboards/doio/kb09/keymaps/default/keymap.c (hunk 6 above)
Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/default/keymap.c]: @drashna Hi， I‘ve replaced it. Thank you!

Comment (DOIO2022) [reply keyboards/doio/kb09/keymaps/default/keymap.c]: @drashna Hi,drashna,is there any problems? please reviewed it, thank you very much!

**Review by drashna — APPROVED** [review]

**Pull request merged**